### ISSUE TITLE
Unnatural and long pause, or silence, any suggestions?

### ISSUE THREAD

**Issue by jeffDebug** [SesameAILabs/csm#145]
Hi,

I know that some other users have also reported that the CSM-1B model exhibits long, unnatural pauses and occasional silence during speech generation (https://huggingface.co/sesame/csm-1b/discussions/14). In my experiments, I found this happened quite often. It usually occurred during a conversation of switching 20 turns.

To give a context, I have cleaned up the text by removing most punctuations except single quote sign and $ sign. I have converted all texts to lower cases. That indeed improved significantly the overall quality of the generated audio. But this long pause, or silence, issue persisted. Here are parameters that I used:

topk: 50
temperature: 0.8
max_audio_length: 15_000

How can I get rid of these long pauses, or silent voices? Is there any specific techniques or tricks that I can apply, via fine-tuning, to improve the prosody of this CSM-1B model? Any ideas, hints or suggestions are most welcomed!

Thanks!

**Comment by ZackHodari** [comment]
The base model has some instability like you've described. Fine-tuning can be complex but is a good way to fix this behavior. You might want to experiment with prompting first though

**Comment by jeffDebug** [comment]
@ZackHodari  Thank you so much for the response. I am wondering, in case you know any specific loss function that is especially good for fine-tuning the natural pause and prosody?

**Comment by ZackHodari** [comment]
standard xent

**Comment by jeffDebug** [comment]
@ZackHodari  Thanks!

Another question, on the hf/official fine-tuing & training guide:
https://huggingface.co/sesame/csm-1b#fine-tuning--training-%F0%9F%93%89

In the code snippet, regarding the training data prep/format:

`ds = load_dataset("eustlb/dailytalk-conversations-grouped", split="train")`

How would the audio-text pairs dataset be prepared for this fine-tuning setup? Are they usually prepared such that many audio (e.g. .wav) files and the corresponding text (.txt) files stored under the same folder, or some other way?

**Comment by ZackHodari** [comment]
I don't know, but you can checkout the example
https://huggingface.co/datasets/eustlb/dailytalk-conversations-grouped/viewer/default/train

**Comment by jeffDebug** [comment]
@ZackHodari  Thanks!  I noticed that speaker_id is part of the grouped training dataset. Does that mean this fine-tuned model also learns the speaker_id in part of the model (weights)? I mean, once fine-tunied, would it be enough just providing the speaker_id during the inference to generate that speaker's voice?

**Comment by ZackHodari** [comment]
If in your FT data a given speaker_id always refers to a specific speaker then the model can learn to clone that speaker

**Comment by jeffDebug** [comment]
@ZackHodari  Thank you so much!

Regarding the input/target data format for CSM model training (or fine-tuning), could you please comment on this line of code in the LoRA implementation for CSM-1b model?
https://github.com/davidbrowne17/csm-streaming/blob/856c0d4ab86fa243ca11cca98afb110eac3a51f6/lora.py#L464

`target_tokens = torch.cat([text_tokens, audio_tokens], dim=0)`

which is confusing to me. Shouldn't the target/output be only _audio_tokens_ instead? Is it incorrect to include _text_tokens_ in the target/output?

By the way, this LoRA implementation is from the closed issue #128 - "Finetuning just dropped".

**Comment by ZackHodari** [comment]
The model does output 33 dims (audio codebooks + text), but we did not pre-train with a loss on this. While you could FT with it I wouldn't recommend it unless you're trying extended pre-training

**Comment by jeffDebug** [comment]
@ZackHodari  Could you please elaborate a bit further on what loss that you guys pre-trained on, and how the loss was computed?

For your reference, this is how the above lora.py computed the loss:

1. codebook0 loss: 
https://github.com/davidbrowne17/csm-streaming/blob/856c0d4ab86fa243ca11cca98afb110eac3a51f6/lora.py#L635

2. codebooks 1 ... 31 loss:
https://github.com/davidbrowne17/csm-streaming/blob/856c0d4ab86fa243ca11cca98afb110eac3a51f6/lora.py#L645

However, if the text was pre-concatenated to the target tokens, I'm wondering if that could be a bug since the 0-index target tokens are not for the audio codebook. Is the _codebook0_ from the backbone output the _text codebook_, which shouldn't be if I'm not mistaken? My understanding is that all 32 codebooks are for audio tokens, except that the codebook0 is the direct VQ, instead of split RVQ.

**Comment by ZackHodari** [comment]
we used cross-entropy on the audio codebooks

since we amortize the decoder training, only ~2 decoder codebooks are used per backbone output (codebook 0). But we're still just computing cross-entropy with the codebooks that are used. In fine-tuning you probably don't need the amortization, it's mainly good for training efficiency at scale

No idea if there are bugs in the OSS implementation, sorry

**Comment by jeffDebug** [comment]
@ZackHodari two more question to clarify:

1. Is it accurate to say that, _all 32 codebooks in the output_ (codebook0 from backbone, and 1...31 from decoder) are _audio codebooks_?
 
2. I read on Sesame's blog article that, 1/16 of output codebooks were randomly sampled to compute the cross-entropy. For 32 codebooks, that's roghtly 2 codebooks size, which makes sense. What about codebook0 that is the output of backbone? Was the x-entropy always computed for codebook0? So the total cross-entropy would be the _x_entropy(codebook0) + x_entropy(1/16 sample codebooks for 1...31)_?

**Comment by ZackHodari** [comment]
1. yes
2. yes

**Comment by jeffDebug** [comment]
@ZackHodari Thanks for the clarification!

By the way, I have tried fine-tuning to a single-speaker's dataset via the LoRA (#128). The long silence issue didn't get improvement at all (maybe a little worse). The single-speaker's dataset consists of audio-text pairs, each of which is a few seconds long (avg 6 to 8 sec). The whole datasets is about 30 min for audio duration in total.

If you don't mind, could you please elaborate a bit about how you guys fine-tuned the CSM base model for the preview/demo version on Sesame's website?

**Comment by ZackHodari** [comment]
Fine-tuning uses the same cross-entropy loss

Getting this to work requires getting a lot of things right, i.e. training code, data, and hparams. All of these impact each other, so to know what advice to give requires full context

I'd recommend joining the community discord, maybe people on there can provide more detailed input for you :)

**Comment by jeffDebug** [comment]
@ZackHodari Cool. I appreciate it! ~~Could you please point me to which discord community, or the url link, where I can join for discussion?~~ (EDIT: never mind this question, I found it in open issue #47)

PS: a follow-up question about the your fine-tuning dataset. Are you guys just fine-tuned on the collection of the single-speaker's audio-clip and text pairs, each of which is 10 to 15 seconds long? What is the total size/duration of your fine-tuning dataset?
